src/views/Referrals/JoinReferralCode.js

```diff
@@ -0,0 +1,151 @@
+import { useEffect, useRef, useState } from "react";
+import { encodeReferralCode, setTraderReferralCodeByUser, validateReferralCodeExists } from "../../Api/referrals";
+import { REFERRAL_CODE_REGEX } from "./ReferralsHelper";
+import { useDebounce } from "../../Helpers";
+import { useWeb3React } from "@web3-react/core";
+
+function JoinReferralCode({ setPendingTxns, pendingTxns, active, connectWallet }) {
+  return (
+    <div className="referral-card section-center mt-medium">
+      <h2 className="title">Enter Referral Code</h2>
+      <p className="sub-title">Please input a referral code to benefit from fee discounts.</p>
+      <div className="card-action">
+        {active ? (
+          <JoinReferralCodeForm setPendingTxns={setPendingTxns} pendingTxns={pendingTxns} />
+        ) : (
+          <button className="App-cta Exchange-swap-button" type="submit" onClick={connectWallet}>
+            Connect Wallet
+          </button>
+        )}
+      </div>
+    </div>
+  );
+}
+
+export function JoinReferralCodeForm({
+  setPendingTxns,
+  pendingTxns,
+  callAfterSuccess,
+  userReferralCodeString = "",
+  type = "join",
+}) {
+  const { account, library, chainId } = useWeb3React();
+  const [referralCode, setReferralCode] = useState("");
+  const inputRef = useRef("");
+  const [isValidating, setIsValidating] = useState(false);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const [referralCodeExists, setReferralCodeExists] = useState(true);
+  const debouncedReferralCode = useDebounce(referralCode, 300);
+
+  function getPrimaryText() {
+    const isEdit = type === "edit";
+    if (isEdit && debouncedReferralCode === userReferralCodeString) {
+      return "Same as current active code";
+    }
+    if (isEdit && isSubmitting) {
+      return "Updating...";
+    }
+
+    if (isSubmitting) {
+      return "Adding...";
+    }
+    if (debouncedReferralCode === "") {
+      return "Enter Referral Code";
+    }
+    if (isValidating) {
+      return `Checking code...`;
+    }
+    if (!referralCodeExists) {
+      return `Referral Code does not exist`;
+    }
+
+    return isEdit ? "Update" : "Submit";
+  }
+  function isPrimaryEnabled() {
+    if (
+      debouncedReferralCode === "" ||
+      isSubmitting ||
+      isValidating ||
+      !referralCodeExists ||
+      debouncedReferralCode === userReferralCodeString
+    ) {
+      return false;
+    }
+    return true;
+  }
+
+  async function handleSubmit(event) {
+    const isEdit = type === "edit";
+    event.preventDefault();
+    setIsSubmitting(true);
+    const referralCodeHex = encodeReferralCode(referralCode);
```

Comment:
i guess we discussed that already – i was thinking about moving all encoding/decoding logic to api/referrals so views should not know about it and handle it (in general i want to keep less logic inside our components/views, because sometimes they're becoming too large and complex).
is it difficult to change, what do you think?